File: animlaboratory/src/main/java/com/skx/tomike/animlaboratory/activity/PropertyAnimatorActivity.java
package com.skx.tomike.animlaboratory.activity;

import android.animation.ObjectAnimator;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationSet;
import android.view.animation.ScaleAnimation;
import android.widget.ImageView;

import com.skx.tomike.animlaboratory.R;

/**
 * 属性动画事例
 */
public class PropertyAnimatorActivity extends AppCompatActivity {
    //    TextView textView_object;
//    TextView textView_value;
    ImageView imageView_animator;
    ImageView imageView2_animator;
//    RelativeLayout container;
//    ViewPropertyHelper textViewTest;
//    ViewPropertyHelper layoutViewProperty;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_property_animatior);
        initializeView();
        refreshView();
        installListener();
    }

    private void initializeView() {
//        textView_object = (TextView) findViewById(R.id.textView_object);
//        textView_value = (TextView) findViewById(R.id.textView_value);
//        container = (RelativeLayout) findViewById(R.id.container);
        imageView_animator = findViewById(R.id.imageView_animator);
        imageView2_animator = findViewById(R.id.imageView2_animator);
    }

    private void refreshView() {
//        textViewTest = new ViewPropertyHelper(textView_value);
//        layoutViewProperty = new ViewPropertyHelper(container);
    }

    private void installListener() {
//        textView_object.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                ObjectAnimator.ofFloat(textView_object, "rotationX", 0.0f, 360.0f).setDuration(500).start();
//            }
//        });
//        textView_value.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                ValueAnimator valueAnimator = ValueAnimator.ofInt(300, 500);
//                valueAnimator.setDuration(500);
//                valueAnimator.setInterpolator(new LinearInterpolator());
//                valueAnimator.setTarget(textViewTest);
//                valueAnimator.start();
//                valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
//                    @Override
//                    public void onAnimationUpdate(ValueAnimator animation) {
//                        Integer animatedValue = (Integer) animation.getAnimatedValue();
//                        textViewTest.setHeight(animatedValue);
//                    }
//                });
//            }
//        });
//        container.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                ValueAnimator valueAnimator = ValueAnimator.ofInt(48, 210);
//                valueAnimator.setDuration(500);
//                valueAnimator.setInterpolator(new LinearInterpolator());
//                valueAnimator.setTarget(layoutViewProperty);
//                valueAnimator.start();
//                valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
//                    @Override
//                    public void onAnimationUpdate(ValueAnimator animation) {
//                        Integer animatedValue = (Integer) animation.getAnimatedValue();
//                        layoutViewProperty.setPadding(animatedValue);
//                    }
//                });
//            }
//        });

        imageView_animator.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                isFavorite = !isFavorite;
                if (!isFavorite) {
                    AnimationSet animationSet = new AnimationSet(true);
                    ScaleAnimation scaleAnimation = new ScaleAnimation(1, 0f, 1, 0f,
                            Animation.RELATIVE_TO_SELF, 0.5f,
                            Animation.RELATIVE_TO_SELF, 0.5f);
                    scaleAnimation.setDuration(200);
                    animationSet.addAnimation(scaleAnimation);
                    animationSet.setFillAfter(true); //让其保持动画结束时的状态。
                    imageView_animator.startAnimation(animationSet);
                } else {
                    AnimationSet animationSet = new AnimationSet(true);
                    ScaleAnimation scaleAnimation = new ScaleAnimation(0, 1f, 0, 1f,
                            Animation.RELATIVE_TO_SELF, 0.5f,
                            Animation.RELATIVE_TO_SELF, 0.5f);
                    scaleAnimation.setDuration(200);
                    animationSet.addAnimation(scaleAnimation);
                    animationSet.setFillAfter(true); //让其保持动画结束时的状态。
                    imageView_animator.startAnimation(animationSet);
                }

            }
        });

        imageView2_animator.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ObjectAnimator.ofFloat(imageView2_animator, "rotationY", 0.0f, 180.0f).setDuration(500).start();
            }
        });
    }

    private boolean isFavorite = true;
}

File: tank/src/main/java/com/skx/tomike/tank/animation/activity/PropertyAnimatorActivity.kt
package com.skx.tomike.tank.animation.activity

import android.animation.ValueAnimator
import android.os.Bundle
import android.widget.EditText
import android.widget.ImageView
import android.widget.RadioGroup
import android.widget.TextView
import androidx.core.widget.addTextChangedListener
import com.alibaba.android.arouter.facade.annotation.Route
import com.skx.common.base.BaseViewModel
import com.skx.common.base.SkxBaseActivity
import com.skx.common.base.TitleConfig
import com.skx.tomike.tank.R
import com.skx.tomike.tank.ROUTE_PATH_PROPERTY

/**
 * 属性动画事例
 */
@Route(path = ROUTE_PATH_PROPERTY)
class PropertyAnimatorActivity : SkxBaseActivity<BaseViewModel<*>>() {

    private val ivTarget: ImageView by lazy {
        findViewById(R.id.iv_propertyAnimator_target)
    }
    private val tvStartBtn: TextView by lazy {
        findViewById(R.id.iv_propertyAnimator_start)
    }

    private var mStartValue: Float = 0.0f
    private var mEndValue: Float = 0.0f
    private var mPivotX: Float = 0.0f
    private var mPivotY: Float = 0.0f
    private var mDuration: Long = 0
    private var mType: Int = 2


    override fun configHeaderTitle(): TitleConfig {
        return TitleConfig.Builder().setTitleText("属性动画").create()
    }

    override fun initParams() {}

    override fun layoutId(): Int {
        return R.layout.activity_property_animatior
    }

    override fun initView() {
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        tvStartBtn.setOnClickListener {
            onAnimStart()
        }
        findViewById<EditText>(R.id.et_propertyAnimator_start).addTextChangedListener {
            mStartValue = if (it != null && "" != it.toString()) {
                java.lang.Float.parseFloat(it.toString())
            } else {
                0.0f
            }
        }
        findViewById<EditText>(R.id.et_propertyAnimator_end).addTextChangedListener {
            mEndValue = if (it != null && "" != it.toString()) {
                java.lang.Float.parseFloat(it.toString())
            } else {
                0.0f
            }
        }
        findViewById<EditText>(R.id.et_propertyAnimator_pivotX).addTextChangedListener {
            mPivotX = if (it != null && "" != it.toString()) {
                java.lang.Float.parseFloat(it.toString())
            } else {
                0.0f
            }
        }
        findViewById<EditText>(R.id.et_propertyAnimator_pivotY).addTextChangedListener {
            mPivotY = if (it != null && "" != it.toString()) {
                java.lang.Float.parseFloat(it.toString())
            } else {
                0.0f
            }
        }
        findViewById<EditText>(R.id.et_propertyAnimator_duration).addTextChangedListener {
            mDuration = if (it != null && "" != it.toString()) {
                java.lang.Long.parseLong(it.toString())
            } else {
                0
            }
        }

        val mRgOptions = findViewById<RadioGroup>(R.id.rg_propertyAnimator_options)
        mRgOptions.check(R.id.tv_propertyAnimator_scan)
        mRgOptions.setOnCheckedChangeListener { group, checkedId ->
            when (checkedId) {
                R.id.tv_propertyAnimator_rotate -> {
                    mType = 10
                }
                R.id.tv_propertyAnimator_rotateX -> {
                    mType = 11
                }
                R.id.tv_propertyAnimator_rotateY -> {
                    mType = 12
                }
                R.id.tv_propertyAnimator_scan -> {
                    mType = 20
                }
                R.id.tv_propertyAnimator_scanX -> {
                    mType = 21
                }
                R.id.tv_propertyAnimator_scanY -> {
                    mType = 22
                }
                R.id.tv_propertyAnimator_translation -> {
                    mType = 30
                }
                R.id.tv_propertyAnimator_translationX -> {
                    mType = 31
                }
                R.id.tv_propertyAnimator_translationY -> {
                    mType = 32
                }
                R.id.tv_propertyAnimator_alpha -> {
                    mType = 40
                }
            }
        }

    }

    private fun onAnimStart() {
        ivTarget.pivotX = mPivotX
        ivTarget.pivotY = mPivotY
        val valueAnimator = ValueAnimator.ofFloat(mStartValue, mEndValue).setDuration(mDuration)
//        valueAnimator.repeatCount = 1
        valueAnimator.setTarget(ivTarget)
        valueAnimator.addUpdateListener {
            when (mType) {
                10 -> {
                    ivTarget.rotationX = it.animatedValue as Float
                }
                11 -> {
                    ivTarget.rotationY = it.animatedValue as Float
                }
                12 -> {
                    ivTarget.rotationY = it.animatedValue as Float
                }
                20 -> {
                    ivTarget.scaleX = it.animatedValue as Float
                    ivTarget.scaleY = it.animatedValue as Float
                }
                21 -> {
                    ivTarget.scaleX = it.animatedValue as Float
                }
                22 -> {
                    ivTarget.scaleY = it.animatedValue as Float
                }
                30 -> {
                    ivTarget.translationX = it.animatedValue as Float
                    ivTarget.translationY = it.animatedValue as Float
                }
                31 -> {
                    ivTarget.translationX = it.animatedValue as Float
                }
                32 -> {
                    ivTarget.translationY = it.animatedValue as Float
                }
                40 -> {
                    ivTarget.alpha = it.animatedValue as Float
                }
            }

        }
        valueAnimator.start()
    }
}